File: 860-design-circular-queue/design-circular-queue.py
class LinkedList:
    def __init__(self, value, nextNode = None):
        self.val = value
        self.next = nextNode

class MyCircularQueue:

    def __init__(self, k: int):
        self.capacity = k
        self.count = 0
        self.head = None
        self.tail = None
        

    def enQueue(self, value: int) -> bool:
        if self.count >= self.capacity:
            return False

        if self.count == 0:
            new_node = ListNode(value)
            self.head = new_node
            self.tail = self.head
        else:
            new_node = ListNode(value)
            self.tail.next = new_node
            self.tail = new_node
        self.count += 1
        return True
        

    def deQueue(self) -> bool:
        if self.count == 0:
            return False
        self.head = self.head.next
        self.count -= 1
        return True
        

    def Front(self) -> int:

        if self.count == 0:
            return -1
        return self.head.val
        

    def Rear(self) -> int:

        if self.count == 0:
            return -1
        return self.tail.val
        

    def isEmpty(self) -> bool:

        return self.count == 0
        

    def isFull(self) -> bool:

        return self.count == self.capacity
        


# Your MyCircularQueue object will be instantiated and called as such:
# obj = MyCircularQueue(k)
# param_1 = obj.enQueue(value)
# param_2 = obj.deQueue()
# param_3 = obj.Front()
# param_4 = obj.Rear()
# param_5 = obj.isEmpty()
# param_6 = obj.isFull()

File: 860-design-circular-queue/design-circular-queue.java
class ListNode{

    int val;
    ListNode next;

    public ListNode(int val) {
        
        this.val = val;
        this.next = null;

    }
}

class MyCircularQueue {
    
    int count;
    int capacity;
    ListNode head;
    ListNode tail;

    public MyCircularQueue(int k) {
        this.capacity = k;
        this.count = 0;
        this.head = null;
        this.tail = null;
        
    }
    
    public boolean enQueue(int value) {

        if (this.count >= this.capacity) {
            return false;
        }

        ListNode newNode = new ListNode(value);

        if (count == 0) {
            head = newNode;
            tail = newNode;
        } else {
            tail.next = newNode;
            tail = newNode;
        }

        count++;
        return true;
        
    }
    
    public boolean deQueue() {
        if (count == 0) {
            return false;
        }

        head = head.next;
        count--;

        if (count == 0) {
            tail = null;
        }

        return true;
        
    }
    
    public int Front() {
        if (count == 0) {
            return -1;
        }
        return head.val;
    }

    public int Rear() {
        if (count == 0) {
            return -1;
        }
        return tail.val;
    }

    public boolean isEmpty() {
        return count == 0;
    }

    public boolean isFull() {
        return count == capacity;
    }
}

/**
 * Your MyCircularQueue object will be instantiated and called as such:
 * MyCircularQueue obj = new MyCircularQueue(k);
 * boolean param_1 = obj.enQueue(value);
 * boolean param_2 = obj.deQueue();
 * int param_3 = obj.Front();
 * int param_4 = obj.Rear();
 * boolean param_5 = obj.isEmpty();
 * boolean param_6 = obj.isFull();
 */